### PR TITLE
Switched to a more performant locking library

### PR DESCRIPTION
**Pull request by MarkCiliaVincenti**
Disclaimer: I am the author of the new library.

Public benchmarks running on GitHub show AsyncNonKeyedLocker to be considerably more performant; in tests NeoSmart.AsyncLock took 9.71x the time and allocated 7.42x the memory.

https://github.com/MarkCiliaVincenti/AsyncNonKeyedLockBenchmarks/actions/runs/7526873065/job/20485876144

### CODE REVIEW

File: LenovoLegionToolkit.WPF/Pages/AboutPage.xaml [hunk 1]
```diff
@@ -65,7 +65,7 @@
 
             <StackPanel Grid.Column="0" Margin="0,0,32,0">
                 <wpfui:Hyperlink Content="Autofac" NavigateUri="https://github.com/autofac/Autofac" />
-                <wpfui:Hyperlink Content="AsyncLock" NavigateUri="https://github.com/neosmart/AsyncLock" />
+                <wpfui:Hyperlink Content="AsyncNonKeyedLocker" NavigateUri="https://github.com/neosmart/AsyncNonKeyedLocker" />
```
Comment (BartoszCichecki) [inline LenovoLegionToolkit.WPF/Pages/AboutPage.xaml:68]: Looks like this link is not correct ;)

**Review by BartoszCichecki — CHANGES_REQUESTED** [review]

File: LenovoLegionToolkit.WPF/Pages/AboutPage.xaml (hunk 1 above)
Comment (MarkCiliaVincenti) [reply LenovoLegionToolkit.WPF/Pages/AboutPage.xaml]: Apologies, fixed.

**Review by BartoszCichecki — APPROVED** [review]

**Comment by BartoszCichecki** [comment]
Just FYI: I reverted changes made in this PR. The behavior of AsyncNonKeyedLocker is slightly different from NeoSmart.AsyncLock, especially when it comes to re-entrance into the same lock.

Example:

```
AsyncNonKeyedLocker lock = new();

public async Task Method1()
{
    using (await lock.LockAsync().ConfigureAwait(false))
    {
        // do some work
        await Method2().ConfigureAwait(false);
    }
}

public async Task Method2()
{
    using (await lock.LockAsync().ConfigureAwait(false))
    {
        // do some work
    }
}
````

In cases like this there is a deadlock that occurs while using your library, while NeoSmart's works properly. Once this issue is resolved, I am more than happy to revist this, even though I don't expect any significant impact on the whole app.

**Comment by MarkCiliaVincenti** [comment]
@BartoszCichecki that is correct, I should make it clear in the library that the locks are not re-entrant.

I might make an attempt at making a re-entrant version one day, however I've so far held back because of the complexities and the problems that could arise under specific conditions, leading to some calling [re-entrant locks impossible](https://itnext.io/reentrant-recursive-async-lock-is-impossible-in-c-e9593f4aa38a). Usually I would advise refactoring code to avoid the need for re-entrancy. It would most likely be more performant anyway.

**Comment by MarkCiliaVincenti** [comment]
If you're interested I could refactor the code in LenovoLegionToolkit to not require re-entrancy and then you could re-apply this PR.

**Comment by MarkCiliaVincenti** [comment]
Some more interesting reading with regards to lock re-entrancy in async code.

https://stackoverflow.com/questions/77834314/when-to-take-await-a-semaphore-as-local-as-possible
https://blog.stephencleary.com/2013/04/recursive-re-entrant-locks.html

**Pull request merged**